SMPL server

WYK96 opened the pull request:
Modified SMPL server so that it is compatible with the bridge server. See commit history for more details.

WYK96 committed:
[bug] Fix docker image startup issue. The following error occured when running docker image:'docker: No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'', which can be solved by adding argument: '--runtime=nvidia'

WYK96 committed:
1.[improvement] Rewrote the SMPL client as a class for better life-cycle management
2.[improvement] Removed dependency for xrmocap. The action that loads the uploaded body motion and gets the metadata(e.g., frame number), has been moved to the server side.

WYK96 committed:
1.[bug] Using 'emit' to send response to the client will not guarantee that the client has  received the response and executed the callback functions before server handles the next request. We use 'socketio.call' instead at the client side to makes sure that the server will handle subsequent requests only after the current request has been processed and the client received response. This is crucial since we need the inferenced body meshes to be temporally coherent.
2.[bug] The socketio limits the maximum allowed payload to be 1MB by default, which is not enough for a long sequence. We added a configurable parameter 'max_http_buffer_size' to control the server's max payload.

WYK96 committed:
1.[feature] add an option 'max_http_buffer_size' to adjust the server's payload, which is 128MB by default.

LazyBusyYang commented on tools/clients/smpl_verts_client.py:
```diff
@@ -1,69 +1,106 @@
+        server_ip: str = '127.0.0.1',
+        server_port: int = 8376,
+        resp_type: str = 'bytes',
+        log_path: Optional[str] = None
```
To get rid of xrprimer, we can have two files for this client:
1. smpl_vert_client.py
```python
# do not import xrprimer/xrmocap in this file
# but import python built-in packages
import logging
class XRMocapSMPLClient:
    def __init__(
        self,
        server_ip: str = '127.0.0.1',
        server_port: int = 8376,
        resp_type: str = 'bytes',
        logger: logging.logger
        ...
```
2. tools/clients/run_smpl_verts_client.py
```python
# it is OK to import xrprimer in this tool
from ..... import XRMocapSMPLClient
logger = setup_logger(...)
client = XRMocapSMPLClient(..., logger=logger)
```

WYK96 replied:
Since the bridge server is started via subprocess, it is hard to maintain the logger throughout the life cycle of the backend and bridge server. Hence, the logger is replaced by console output in the SMPL client.

LazyBusyYang replied:
We have some examples help subprocess writes the same log file in XRMoDaMa. No worry, I'll fix it later.

LazyBusyYang commented on tools/clients/smpl_verts_client.py:
```diff
@@ -1,69 +1,106 @@
+
+        return t
+    
+    def __parse_upload_response(self, data):
```
Not a good PEP8 naming.
PEP8: _single_leading_underscore: weak “internal use” indicator. E.g. from M import * does not import objects whose names start with an underscore.

see [ref](https://peps.python.org/pep-0008/#descriptive-naming-styles) here.

WYK96 replied:
Replaced double leading underscore to single leading underscore.

LazyBusyYang commented on xrmocap/service/smpl_verts_service.py:
```diff
@@ -53,6 +53,7 @@ def __init__(self,
                  device: Union[torch.device, str] = 'cuda',
                  host: str = '0.0.0.0',
                  port: int = 29091,
+                 max_http_buffer_size: int = 128 * 1024 * 1024,
```
For a default value, 134217728 is faster, and we need a docstring at L92, introducing it's 128MB and what is it for.

WYK96 replied:
Consider a case that we want to set max_http_buffer_size from 128MB to 64MB. If the buffer size is expressed as 128 * 1024 * 1024, the user is able to achieve the goal by simply modifying 128 to 64, otherwise he/she need to manually compute the 64* 1024 * 1024 = 67,108,864 which is not reading-friendly. 

134217728 is faster whereas I do not regard  128 * 1024 * 1024 as a computational overhead since it's only an integer multiply opertation. I suggest using 128 * 1024 * 1024 to maintain the readability.

LazyBusyYang replied:
OK. But please do not modify the default value in function definition, pass another new value is good.

LazyBusyYang commented on xrmocap/service/smpl_verts_service.py:
```diff
@@ -118,14 +120,8 @@ def __init__(self,
-            handler=self.forward_body_model,
-        )
+        
+        self.call_backs()
```
We have two styles for binding: one is decorator `@self.socketio.on`, the other is function call `self.socketio.on_event()`. Why don't we use only one of them?

WYK96 replied:
self.call_backs() is used to register the events that the server is going to listen for. Inside the self.call_backs() function body are event functions decorated by @self.socketio.on. If the event functions are directly defined and implemented in the __init__, the __init__ will look really long(See https://github.com/ParisNeo/lollms/blob/2c0d9af459150b4d6901ecdf48f57866f9eab89b/lollms/apps/server/__init__.py#L98 for similar usage). Besides, it seems that socketio does not provide an option to register routes by socketio.on(mesage, handler). Did I miss something?

WYK96 replied:
UPDATE: I found that self.socketio.on_event can be used to register routes. Sorry for the misconception.

LazyBusyYang commented on xrmocap/service/smpl_verts_service.py:
```diff
@@ -168,122 +164,156 @@ def on_disconnect(self) -> None:
+            session['smpl_gender'] = smpl_data.get_gender()
+            session['smpl_data'] = smpl_data
+            session['body_model'] = body_model
+            session['faces'] = body_model.faces
```
Is `.cpu().numpy().tolist/tobytes()` needed here?

LazyBusyYang replied:
1. `ndarray` is better than `torch.Tensor` for arguments and attributes
2. it is easy to get faces from `session['body_model']`, there's no need to keep `session['faces']`

WYK96 replied:
Removed session['faces'].

LazyBusyYang commented on xrmocap/service/smpl_verts_service.py:
```diff
@@ -168,122 +164,156 @@ def on_disconnect(self) -> None:
-                emit('forward_response', verts_bytes)
+                faces = np.array(faces, dtype=np.int32)
+                faces_bytes = faces.tobytes()
+
```
`enable_gzip` is also possible here.

WYK96 replied:
Added enable_gzip for compressing faces_bytes.

LazyBusyYang commented on xrmocap/service/smpl_verts_service.py:
```diff
@@ -168,122 +164,156 @@ def on_disconnect(self) -> None:
+            faces = session['faces']
+            # check if data and args are valid
+            success = True
+            if faces is None:
+                error_msg = f'Failed to get body faces.'
+                success = False
+
+            if success is False:
```
The `if` in L298 seems unnecessary.

WYK96 replied:
Merged ```if``` in L298 with L294.

LazyBusyYang commented on xrmocap/service/smpl_verts_service.py:
```diff
@@ -168,122 +164,156 @@ def on_disconnect(self) -> None:
+        @self.socketio.on('get_faces')
+        def get_faces() -> dict:
+            resp_dict = dict()
+            faces = session['faces']
```
faces is included in the body model, it is not necessary to save it in session again. 
```suggestion
            body_model = session['body_model']
            if body_model is None:
                # log and respond
            else:
                faces_np = body_model.faces.cpu().numpy()
```

WYK96 replied:
1. Removed ```session['faces']```.
2. ```body_model.faces``` is already a ndarray, therefore, ```faces.cpu().numpy()``` is no longer needed.

LazyBusyYang commented on xrmocap/service/smpl_verts_service.py:
```diff
@@ -168,122 +164,156 @@ def on_disconnect(self) -> None:
+        
+        @self.socketio.on('get_faces')
+        def get_faces() -> dict:
+            resp_dict = dict()
```
Need docstring. This [plugin](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) may help you.

WYK96 replied:
Added docstring. Thanks for the suggestion.

LazyBusyYang commented on tools/clients/smpl_verts_client.py:
```diff
@@ -74,59 +111,20 @@ def on_forward_response(data):
+        return verts.tolist()
+
+    def forward(self, frame_idx: int):
+        self.logger.info(f'Requesting frame {frame_idx}...')
```
`logger.info` is too noisy if we call it every frame.

WYK96 replied:
It was used for the debug purpose. Now it has been removed.

LazyBusyYang commented on tools/clients/smpl_verts_client.py:
```diff
@@ -1,69 +1,106 @@
 
-from xrmocap.data_structure.body_model import auto_load_smpl_data
+class XRMocapSMPLClientActionsEnum(str, Enum):
+    UPLOAD = 'upload'
```
Enum maps str to int and makes int easier to read. If the target is a string, cannot see any enhance here.

WYK96 replied:
From ChatGPT:

Using `str` enums in Python can be quite useful depending on the situation. Enums provide a way to define a set of named values that are treated as constants, which can improve the readability and maintainability of your code. 

When you define an `enum` in Python, you create a new class with a set of pre-defined values. These values can be compared using the `==` operator, and they can also be iterated over like any other iterable object in Python. Here's an example:

```python
from enum import Enum

class Color(Enum):
    RED = 'red'
    GREEN = 'green'
    BLUE = 'blue'

print(Color.RED)  # Color.RED
print(Color.RED.value)  # 'red'
print(list(Color))  # [<Color.RED: 'red'>, <Color.GREEN: 'green'>, <Color.BLUE: 'blue'>]
```

In this example, we define a new `enum` called `Color` with three values: `RED`, `GREEN`, and `BLUE`, each with a corresponding string value. We can access the enum values using the dot notation, and we can also access the string value of each enum value using the `value` attribute.

One benefit of using `str` enums is that they can help catch errors at compile time. For example, if you have a function that takes a `Color` enum as an argument, you can be sure that the argument will always be one of the predefined values, rather than a random string.

Overall, using `str` enums can make your code more readable, maintainable, and less prone to errors.

---------------------------------------------------------------------------------------------------------------------

My summary of benefits from using str enums:
1. Improve the readability and maintenance.
2. Prevent incorrect events(e.g., the user typed a random str)

WYK96 committed:
1.[improvement] add docstring
    2.[new feature] add support for smplx
    3.[improvement] remove dependency for xrprimer

WYK96 committed:
changed server default port to 18800

WYK96 committed:
1.[improvement] improved routes registration
2.[improvement] add docstring
